crates/hyperdrive-math/src/long/max.rs

```diff
@@ -2,7 +2,623 @@ use ethers::types::I256;
 use fixed_point::FixedPoint;
 use fixed_point_macros::{fixed, int256};
 
-use crate::{State, YieldSpace};
+use crate::{
+    calculate_open_long, open_long_curve_fees, open_long_governance_fee,
+    yield_space::{get_spot_price, k_down},
+    State, YieldSpace,
+};
+
+/// Gets the pool's max spot price.
+///
+/// Hyperdrive has assertions to ensure that traders don't purchase bonds at
+/// negative interest rates. The maximum spot price that longs can push the
+/// market to is given by:
+///
+/// $$
+/// p_max = \frac{1 - \phi_f}{1 + \phi_c * \left( p_0^{-1} - 1 \right) * \left( \phi_f - 1 \right)}
+/// $$
+pub fn get_max_spot_price(
+    ze: FixedPoint,
+    y: FixedPoint,
+    mu: FixedPoint,
+    t: FixedPoint,
+    flat_fee: FixedPoint,
+    curve_fee: FixedPoint,
+) -> FixedPoint {
+    (fixed!(1e18) - flat_fee)
+        / (fixed!(1e18)
+            + curve_fee.mul_up(fixed!(1e18).div_up(get_spot_price(ze, y, mu, t)) - fixed!(1e18)))
+        .mul_up(fixed!(1e18) - flat_fee)
+}
+
+/// Gets the pool's solvency.
+///
+/// $$
+/// s = z - \tfrac{exposure}{c} - z_min
+/// $$
+pub fn get_solvency(
+    z: FixedPoint,
+    c: FixedPoint,
+    z_min: FixedPoint,
+    long_exposure: FixedPoint,
+) -> FixedPoint {
+    z - long_exposure / c - z_min
+}
+
+/// Gets the max long that can be opened given a budget.
+///
+/// We start by calculating the long that brings the pool's spot price to 1.
+/// If we are solvent at this point, then we're done. Otherwise, we approach
+/// the max long iteratively using Newton's method.
+pub fn get_max_long<F: Into<FixedPoint>, I: Into<I256>>(
+    zeta: I256,
+    ze: FixedPoint,
+    z: FixedPoint,
+    y: FixedPoint,
+    c: FixedPoint,
+    mu: FixedPoint,
+    t: FixedPoint,
+    z_min: FixedPoint,
+    time_stretch: FixedPoint,
+    c_initial: FixedPoint,
+    flat_fee: FixedPoint,
+    curve_fee: FixedPoint,
+    governance_lp_fee: FixedPoint,
+    long_exposure: FixedPoint,
+    budget: F,
+    checkpoint_exposure: I,
+    maybe_max_iterations: Option<usize>,
```

Comment:
I see what you mean now, these are pretty big and gnarly. I do think that some structs would clean this up. Having a struct that contains all of the yield space stuff seems like it would get us pretty far.

Reply:
I'll leave it to you if you want to make this change. It would be good to get a review from @dpaiton or @slundqui since they are the people who requested this and are the most likely consumers.